### DISCUSSION TITLE
C++ unique object & Python weak reference

### DISCUSSION THREAD

**Discussion by lianera** [wjakob/nanobind#409] [General]
Is it possible to manage ownership using std::unique_ptr in C++, while obtaining only weak references in Python? When the unique_ptr is destroyed, the weak references in Python should also become invalid simultaneously. 
Recently, I implemented something similar, an intrusive C++ UniqueObject that holds a pointer to additional block for storing the reference count of weak references. However, when converting weak references to Python, an object needs to be created each time, resulting in additional overhead, and the created weak references are not the same object. 
I'm wondering if it's possible to simply mark the Python object associated with the weak reference as invalid when the UniqueObject is destroyed. Later, when using the weak reference's Python object, an exception could be thrown if the object has been marked invalid.

**Comment by wjakob** [comment]
It's not supported by stock nanobind. You might be able to achieve such features by modifying the library -- however, that is beyond the scope of this support forum.
